system/lib/wasmfs/wasmfs.cpp

```diff
@@ -156,20 +170,55 @@ __wasi_errno_t __wasi_fd_read(__wasi_fd_t fd,
 
     size_t bytesToRead = std::min(size_t(dataLeft), iovs[i].buf_len);
 
-    auto result = lockedFile.read(buf, bytesToRead, offset);
+    auto result = lockedFile.read(buf, bytesToRead, currOffset);
 
     if (result != __WASI_ERRNO_SUCCESS) {
-      *nread = offset - lockedOpenFile.position();
-      lockedOpenFile.position() = offset;
+      *nread = currOffset - lockedOpenFile.position();
+      if (setPosition == overwrite) {
+        lockedOpenFile.position() = currOffset;
+      }
       return result;
     }
-    offset += bytesToRead;
+    currOffset += bytesToRead;
+  }
+  *nread = currOffset - oldOffset;
+  if (setPosition == overwrite) {
+    lockedOpenFile.position() = currOffset;
   }
-  *nread = offset - lockedOpenFile.position();
-  lockedOpenFile.position() = offset;
   return __WASI_ERRNO_SUCCESS;
 }
 
+__wasi_errno_t __wasi_fd_write(__wasi_fd_t fd,
+                               const __wasi_ciovec_t* iovs,
+                               size_t iovs_len,
+                               __wasi_size_t* nwritten) {
+  return writeAtOffset(fd, iovs, iovs_len, NULL, nwritten, Position::overwrite);
```

Comment:
Is this passing `NULL` as the offset? That seems confusing, as NULL is normally used only for pointers. (`nullptr` would also fail to validate here, and is a good reason in general to use `nullptr` over `NULL`)
```suggestion
  return writeAtOffset(fd, iovs, iovs_len, 0, nwritten, Position::overwrite);
```

Reply:
Oh yeah, using a proper optional parameter makes sense to me. It might also be nice to move the `Position` enum to be the first parameter, since it generally controls the behavior of the function.